### ISSUE TITLE
[Bug] Tabs are disappearing briefly (blinking / re-rendering)

### ISSUE THREAD

**Issue by thnee** [piroor/treestyletab#3597]
## Abstract

The GUI produces a type of "blinking effect" during normal usage.
The tabs disappear briefly, the sidebar panel is blank, and then the tabs are visible again.

## Related issue

Let me describe a similar issue, just to clarify that I think this is a different issue.
Sometimes, it seems like the tabs are reloaded from storage.
The tabs disappear, and there is a loading animation which takes a couple of seconds, and then all the tabs are loaded up again.
(It seems like the code is reloading tabs from disk into memory. It seems like this is an intentional behavior. Although I am not sure why this happens).

## Expected result

GUI remains visible at all times, like every other normal GUI program.

## Actual result

The tabs disappear briefly, the body of the sidebar panel is blank, and then the tabs are visible again.

This happens every time when switching to the Firefox window from another app window (with Alt-Tab).

Most of the time it happens very quickly, like 20 milliseconds. Then it is noticeable, but only slightly annoying.
But sometimes it takes a lot longer, like 500 milliseconds. Then it is a lot more annoying.

It looks like it is re-rendering the GUI content.
Nothing else is changing.
\- There is no loading animation.
\- The tab structure/order is not changing.
\- The scrollbar position is not changing.

The only content that is blinking are the tabs inside the sidebar panel.
Everything else is behaving in a normal expected way, and is not blinking.
\- The TST sidebar panel background and border is not blinking.
\- The color is not changing. There is no white flash when using dark theme.
\- Other parts of the Firefox Window are not blinking.
\- Other GUI programs on the system are not blinking.

## Environment

- Platform (OS): AMD laptop with Windows 11 Pro 23H2
- Version of Firefox: 128.0.3 (64-bit)
- Version (or revision) of Tree Style Tab: 4.0.20 (2024.6.28)

This issue happens on both my Intel desktop with Windows 10, and on my AMD laptop with Windows 11.
But the problem is generally worse on my AMD laptop, this is where the time is usually longer, like 500 milliseconds.

This issue happens on all displays, such as internal laptop display, and external USB-C display, and external DisplayPort display.

**Comment by piroor** [comment]
> Most of the time it happens very quickly, like 20 milliseconds. Then it is noticeable, but only slightly annoying.

Does this mean that TST's sidebar panel is always re-rendered or reloaded on switching of browser windows? It is unexpected - I didn't design TST to do that. There looks to be something unknown trigger.

TST is built on two parts: the background part and sidebar panel parts. They track Firefox's tabs in parallel, and sidebar panels are reloaded when TST detects mismatching of their tracking tabs information or detects disconnection between those parts, to synchronize them forcefully. The case taking 500msec looks to be this pattern, and it often happens when there are other tab-related addons or you operate multiple tabs very quickly. Do you use any other addons or very large number tabs?

**Comment by thnee** [comment]
Thank you for the reply!

Yes, it happens every time.

The problem only happens when the windows are maximized.
If I switch from a not-maximized window, the problem does not happen.

If I alt-tab between maximized windows quickly, there is a visible re-render in the tabs in the TST sidebar.
The flicker happens on every single alt-tab, but it is very short, like 20 milliseconds.

If I leave Firefox in the background for some time, like 30 seconds, the re-render is sometimes longer.
Sometimes it is short like 20 milliseconds, and sometimes it is longer like a few hundred milliseconds.
(Background means: open, and maximized, but hidden under other maximized windows).

I do have a lot of tabs open in my main window, and I do have a lot of extensions.
I also have the following in `userChrome.css`.
```
#TabsToolbar
{
    visibility: collapse;
}
```

But I have tested this with all other extensions disabled.
And I have tested with an empty `userChrome.css`.
And I have tested with a new window that only has 2 tabs.
And the the problem still happens, there is no noticeable difference.

**Comment by piroor** [comment]
I've tried with two windows: one is fullscreen. Steps I did:

1. Open two browser windows.
2. Open some tabs in both windows.
3. Show TST's sidebar panel on both windows.
4. Hit F11 key to switch a browser window to the fullscreen mode.
5. Hit Alt-Tab to switch focus between two browser windows.

On this case I've seen flashing sidebar panel on the normal (not fullscreen) window. Interestingly it happens not only with TST. I've confirmed that it happens with another sidebar addon Sidebery. So it looks to be a bug of Firefox itself.

But I still cannot reproduce the problem of unexpected reloading of TST's sidebar panel. Hmm...

**Comment by adderek** [comment]
Several problems are there and several possible solutions:
1. Please, update your graphic driver. If this doesn't help try disabling hardware acceleration in firefox settings. I have observed similar issues several times: if you look closely then each "object" is drawn on the screen (might be too fast for your eye to catch it) and this is why it is blinking.
2. Another issue is a failure in the TST itself. TST disappears for a moment (becoming an huge empty rectangle) and renders the tree again. To troubleshoot your issue you should debug the extension. For me such a problem is with session restore (tab session manager) which tries to restore tabs but those cannot be accessed. **Other causes for same effect could be there.** I have only debugged this tot he symptoms level and code which reports the error, while not being able to find the root cause. Could be some timing issue (it all works in JS and a single thread although asynchronous) and I have normally between 200 and 4000 tabs open (most unloaded/discarded but up to several hundreds could be in ram).
3. If your TST become an permanently empty space instad of a tree - you could try to open and close it again (press F1 twice on a windows/linux PC)
